### ISSUE TITLE
样本数据的格式转换

### ISSUE THREAD

**Issue by zhangzhenyue1** [PaddlePaddle/Paddle#2767]
请问有将libsvm类的label feaid:feavalue feaid:feavalue feaid:feavalue样本数据，转换成paddle的dnn样本输入的函数吗？feaid是稀疏形式的表示。

谢谢

**Comment by Superjom** [comment]
DNN的输入是 float 的向量？ 
直接 

```python
import numpy as np
dim = 300
libsvm_fea = "0 3:1 12:0.5"

def transform_from_libsvm_to_dnn(lib_svm_fea):
    dnn_fea = np.zeros(dim)
    for (id,f) in libsvm_fea.split():
        id = int(id); f = float(f)
        dnn_fea[id] = f
    return dnn_fea

res = transform_from_libsvm_to_dnn(libsvm_fea) 
```

**Comment by zhangzhenyue1** [comment]
是float向量。
必须是稠密的表示方法吗？
dnn_fea = np.zeros(dim) 这句是否可以去掉？

**Comment by zhangzhenyue1** [comment]
现在我们有一个libsvm的训练数据，能否提供一个简单的使用dnn的train代码样例，谢谢！

**Comment by zhangzhenyue1** [comment]
我是按照https://github.com/PaddlePaddle/models/blob/develop/ctr/train.py#L113这个例子修改的，想问下：

dnn_merged_input = layer.data(
  | name='dnn_input',
  | type=xxx)
这里的type怎么写比较好，谢谢！

**Comment by zhangzhenyue1** [comment]
目前代码可以运行，train部分正常，我导入test数据，想算auc的时候，调用infer函数报错。由于没有找到合适的例子，还请帮忙看一下。
输入数据是"0 3:1 12:0.5"这样的格式，代码如下：
`#!/usr/bin/env python
# -*- coding: utf-8 -*-

import argparse
import logging
import paddle.v2 as paddle
from paddle.v2 import layer
from paddle.v2 import data_type as dtype
from data_provider import AvazuDataset

dim=686

parser = argparse.ArgumentParser(description="PaddlePaddle CTR example")
parser.add_argument(
    '--train_data_path',
    type=str,
    required=True,
    help="path of training dataset")
parser.add_argument(
    '--test_data_path',
    type=str,
    required=True,
    help="path of test dataset")
parser.add_argument(
    '--batch_size',
    type=int,
    default=10000,
    help="size of mini-batch (default:10000)")
parser.add_argument(
    '--test_set_size',
    type=int,
    default=10000,
    help="size of the validation dataset(default: 10000)")
parser.add_argument(
    '--num_passes', type=int, default=2, help="number of passes to train")
parser.add_argument(
    '--num_lines_to_detact',
    type=int,
    default=50000,
    help="number of records to detect dataset's meta info")

args = parser.parse_args()

print args.num_passes
print args.train_data_path
print args.test_data_path
print args.batch_size
print args.test_set_size
print args.num_lines_to_detact

dnn_layer_dims = [128, 64, 32, 1]

paddle.init(use_gpu=False, trainer_count=1)

# ==============================================================================
#                    input layers
# ==============================================================================
dnn_merged_input = layer.data(
    name='dnn_input',
    type=paddle.data_type.sparse_value_slot(dim))

click = paddle.layer.data(name='click', type=dtype.dense_vector(1))

# ==============================================================================
#                    network structure
# ==============================================================================
def build_dnn_submodel(dnn_layer_dims):
    dnn_embedding = layer.fc(input=dnn_merged_input, size=dnn_layer_dims[0])
    _input_layer = dnn_embedding
    for i, dim in enumerate(dnn_layer_dims[1:]):
        fc = layer.fc(
            input=_input_layer,
            size=dim,
            act=paddle.activation.Relu(),
            name='dnn-fc-%d' % i)
        _input_layer = fc
    fc = layer.fc(
        input=_input_layer,
        size=1,
        name='output',
        # use sigmoid function to approximate ctr rate, a float value between 0 and 1.
        act=paddle.activation.Sigmoid())
    return fc

output = build_dnn_submodel(dnn_layer_dims)

# ==============================================================================
#                   cost and train period
# ==============================================================================
classification_cost = paddle.layer.multi_binary_label_cross_entropy_cost(
    input=output, label=click)

params = paddle.parameters.create(classification_cost)

optimizer = paddle.optimizer.Momentum(momentum=0.01)

trainer = paddle.trainer.SGD(
    cost=classification_cost, parameters=params, update_equation=optimizer)

dataset = AvazuDataset(
    args.train_data_path, args.test_data_path, n_records_as_test=args.test_set_size)

field_index = {'dnn_input': 0, 'click': 1}

def event_handler(event):
    if isinstance(event, paddle.event.EndIteration):
        num_samples = event.batch_id * args.batch_size
        if event.batch_id % 100 == 0:
            logging.warning("Pass %d, Samples %d, Cost %f, size %d" %
                            (event.pass_id, event.batch_id, event.cost, args.batch_size))

        if event.batch_id % 300 == 0:
            result = trainer.test(
                reader=paddle.batch(dataset.test, batch_size=args.batch_size),
                feeding=field_index)
            logging.warning("Test %d-%d, Cost %f, %s" %
                            (event.pass_id, event.batch_id, result.cost, result.metrics))


trainer.train(
    reader=paddle.batch(
        paddle.reader.shuffle(dataset.train, buf_size=500),
        batch_size=args.batch_size),
    feeding=field_index,
    event_handler=event_handler,
    num_passes=args.num_passes)


infer_dict = copy.copy(field_index)
del infer_dict['click']
prediction = paddle.infer(output, parameters=params, input=[(dataset.test_fea)])

label = (dataset.test_label)

auc_eval = auc_evaluator(prediction, label)
print auc_eval
`

`import sys
import csv
import numpy as np

def transform_from_libsvm_to_dnn(lib_svm_fea):
    dnn_fea = []
    for data in lib_svm_fea.split(' '):
        feas = data.split(':')
        if len(feas)<2:
            continue
        id = int(feas[0]); f = float(feas[1])
        #dnn_fea[id] = f
        dnn_fea.append((id, f))
    return dnn_fea

class AvazuDataset(object):
    '''
    Load AVAZU dataset as train set.
    '''
    TRAIN_MODE = 0
    TEST_MODE = 1

    def __init__(self, train_path, test_path, n_records_as_test=-1):
        self.train_path = train_path
        self.test_path = test_path
        # task model: 0 train, 1 test
        self.mode = 0

    def train(self):
        self.mode = self.TRAIN_MODE
        return self._parse(self.train_path)

    def test(self):
        self.mode = self.TEST_MODE
        return self._parse(self.test_path)

    def _parse(self, path, skip_n_lines=-1, top_n_lines=-1):
        i = 0
        data_lines = open(path).readlines()
        # line example: label feaid:value feaid:value feaid:value
        for line in data_lines:
            i = i + 1
            label = []
            label.append(int(line[0]))
            fea = transform_from_libsvm_to_dnn(line[2:])
            record = [fea, label]
            yield record

    def test_fea(self):
        data_lines = open(self.test_path).readlines()
        for line in data_lines:
            fea = transform_from_libsvm_to_dnn(line[2:])
            yield fea

    def test_label(self):
        data_lines = open(self.test_path).readlines()
        for line in data_lines:
            label = []
            label.append(int(line[0]))
            yield label


if __name__ == '__main__':
    path = 'train.txt'

    filereader = AvazuDataset(path)
    for no, rcd in enumerate(filereader.train()):
        print no, rcd
        if no > 10: break
`

**Comment by zhangzhenyue1** [comment]
[train.txt](https://github.com/PaddlePaddle/Paddle/files/1144227/train.txt)
[data_provider.txt](https://github.com/PaddlePaddle/Paddle/files/1144228/data_provider.txt)

**Comment by Superjom** [comment]
可以试下这个 https://github.com/Superjom/models-1/tree/ctr2/ctr

**Comment by zhangzhenyue1** [comment]
@Superjom 
好的，对于稀疏数据，infer输入的数据格式需要做什么额外处理吗，和训练的数据格式保持一致就可以了？
还有就是infer的结果，和之前输入的样本数据顺序一致吗？

**Comment by Superjom** [comment]
没有label，其他应该一样。 
文档里有提到把

**Comment by zhangzhenyue1** [comment]
我将infer的代码改为：
inferer = paddle..inference.Inference(output_layer=output, parameters=params, )
infer_reader = paddle.batch(dataset.test_fea, batch_size=1000)
for id, batch in enumerate(infer_reader()):
    res = inferrer.infer(input=batch)
最后一句话报错：
res = inferrer.infer(input=batch)
最下面的堆栈信息是TypeError: 'int' object has no attribute '__getitem__'

疑问：
看来是dataset.test_fea这个产出的数据不对。
dataset.train 最后是yield fea, label
dataset.test 最后是yield fea, label
dataset.test_fea 最后是yield fea
前两个函数里面去都是能拿到数据的。我自己把数据打印出来，前两个有label的数据样子是：
([id,value], [id, value], [id, value], [id, value], [label])，
后面一个纯特征的是：
[id,value], [id, value], [id, value], [id, value]

**Comment by zhangzhenyue1** [comment]
![wechatimg12](https://user-images.githubusercontent.com/5304770/28162260-09c68164-67f9-11e7-8dc7-c56bd55691fd.jpeg)

附报错图片

**Comment by zhangzhenyue1** [comment]
@Superjom 
使用了你的代码，在infer的时候报错。
1. infer.py的47行左右，需要加一下args.model_type = ModelType(args.model_type)
2. network_config.py的97行附近，需要加一下if not self.is_infer:

**Comment by Superjom** [comment]
赞，我在pr里修改下。 还没merge进官方的仓库里

**Comment by zhangzhenyue1** [comment]
再提个建议：现在的代码逻辑是每迭代一次要load一次数据。这个可以做成只load一次数据的。

**Comment by zhangzhenyue1** [comment]
小数据集的时候没问题，数据量变大之后，infer报错。
![Uploading WechatIMG13.jpeg…]()

**Comment by zhangzhenyue1** [comment]
看起来是数据问题。做了两个尝试：
1. 换了一种类型的数据
2. 还是用原来数据，不用lr，只用dnn
2种方法都能在内网机器正常运行

**Comment by buptwds** [comment]
请问你的libsvm格式数据怎么跑dnn的，直接跑？还是像ctr例子中那样使用了wide&deep模型
